format/macho/macho.go

```diff
@@ -0,0 +1,851 @@
+	0x1: "SG_HIGHVM",
+	0x2: "SG_FVMLIB",
+	0x4: "SG_NORELOC",
+	0x8: "SG_PROTECTED_VERSION_1",
+}
+
+var sectionTypes = scalar.UToSymStr{
+	0x0:  "S_REGULAR",
+	0x1:  "S_ZEROFILL",
+	0x2:  "S_CSTRING_LITERALS",
+	0x3:  "S_4BYTE_LITERALS",
+	0x4:  "S_8BYTE_LITERALS",
+	0x5:  "S_LITERAL_POINTERS",
+	0x6:  "S_NON_LAZY_SYMBOL_POINTERS",
+	0x7:  "S_LAZY_SYMBOL_POINTERS",
+	0x8:  "S_SYMBOL_STUBS",
+	0x9:  "S_MOD_INIT_FUNC_POINTERS",
+	0xa:  "S_MOD_TERM_FUNC_POINTERS",
+	0xb:  "S_COALESCED",
+	0xc:  "S_GB_ZEROFILL",
+	0xd:  "S_INTERPOSING",
+	0xe:  "S_16BYTE_LITERALS",
+	0xf:  "S_DTRACE_DOF",
+	0x10: "S_LAZY_DYLIB_SYMBOL_POINTERS",
+	0x11: "S_THREAD_LOCAL_REGULAR",
+	0x12: "S_THREAD_LOCAL_ZEROFILL",
+	0x13: "S_THREAD_LOCAL_VARIABLES",
+	0x14: "S_THREAD_LOCAL_VARIABLE_POINTERS",
+	0x15: "S_THREAD_LOCAL_INIT_FUNCTION_POINTERS",
+}
+
+var sectionFlags = map[uint64]string{
+	0x00000100: "S_ATTR_LOC_RELOC",
+	0x00000200: "S_ATTR_EXT_RELOC",
+	0x00000400: "S_ATTR_SOME_INSTRUCTIONS",
+
+	0x02000000: "S_ATTR_DEBUG",
+	0x04000000: "S_ATTR_SELF_MODIFYING_CODE",
+	0x08000000: "S_ATTR_LIVE_SUPPORT",
+	0x10000000: "S_ATTR_NO_DEAD_STRIP",
+	0x20000000: "S_ATTR_STRIP_STATIC_SYMS",
+	0x40000000: "S_ATTR_NO_TOC",
+	0x80000000: "S_ATTR_PURE_INSTRUCTIONS",
+}
+
+func machoDecode(d *decode.D, in interface{}) interface{} {
+	ofileDecode(d)
+	return nil
+}
+
+func ofileDecode(d *decode.D) {
+	var archBits int
+	var cpuType uint64
+	var ncmds uint64
+	magicBuffer := d.U32LE()
+
+	if magicBuffer == MH_MAGIC || magicBuffer == MH_MAGIC_64 {
+		d.Endian = decode.LittleEndian
+		if magicBuffer == MH_MAGIC {
+			archBits = 32
+		} else {
+			archBits = 64
+		}
+	} else if magicBuffer == MH_CIGAM || magicBuffer == MH_CIGAM_64 {
+		d.Endian = decode.BigEndian
+		if magicBuffer == MH_CIGAM {
+			archBits = 32
+		} else {
+			archBits = 64
+		}
+	} else if magicBuffer == FAT_MAGIC {
+		d.Endian = decode.LittleEndian
+		fatParse(d)
+		return
+	} else if magicBuffer == FAT_CIGAM {
+		d.Endian = decode.BigEndian
+		fatParse(d)
+		return
+	} else {
+		// AR files are also valid OFiles but they should be parsed by `-d ar`
+		d.Fatalf("Invalid magic field")
+	}
+
+	d.SeekAbs(0)
+	d.FieldStruct("header", func(d *decode.D) {
+		d.FieldValueS("arch_bits", int64(archBits))
+		magic := d.FieldU32("magic", scalar.Hex, magicSymMapper)
+		d.FieldValueU("bits", uint64(archBits))
+		d.FieldValueStr("endian", endianNames[magic])
+		cpuType = d.FieldU32("cputype", cpuTypes)
+		d.FieldU32("cpusubtype", cpuSubTypes[cpuType])
+		d.FieldU32("filetype", fileTypes)
+		ncmds = d.FieldU32("ncdms")
+		d.FieldU32("sizeofncdms")
+		d.FieldStruct("flags", parseFlags(machHeaderFlags))
+		if archBits == 64 {
+			d.FieldRawLen("reserved", 4*8, d.BitBufIsZero())
+		}
+	})
+	ncmdsIdx := 0
+	d.FieldStructArrayLoop("load_commands", "load_command", func() bool {
+		return ncmdsIdx < int(ncmds)
+	}, func(d *decode.D) {
+		cmd := d.FieldU32("cmd", loadCommands)
+		cmdsize := d.FieldU32("cmdsize")
```

Comment:
I wonder if you could use `d.FramedFn(...)` here if no decoder inside of it needs to dead outside the framed range. Then for example `d.FieldRawLen("unknown", int64((cmdsize-8)*8))` at the end can be `d.FieldRawLen("unknown", d.BitsLeft())`